### PR TITLE
Data parallel tensor

### PR DESCRIPTION
**Pull request by sanketpurandare**
We introduce the DataParalleTensor subclass to enable a simple, eager, easy-to-debug and composable way of expressing data parallel operations in PyTorch.

What is DataParallelTensor?

The DataParallelTensor is a Tensor subclass that internally maintains a list of tensors (one per device) instead of a single tensor. The DataParallelTensor uses __torch_dispatch__ under the hood to execute the aten operations in data parallel mode on all the devices. It automatically lifts the tensor args and kwargs in the operations to DataParallelTensors before execution. Subsequent to execution it wraps the outputs as DataParallelTensors thereby maintaining the flow of execution and freeing the users from associated complexity.
Screen Shot 2022-06-29 at 1 22 55 PM

Why is DataParallelTensor needed?

1. Ease of Debugging: Using DDP is tedious, requires setting up a process group and applying process rank dependent code for debugging thereby requiring expertise. With DataParallelTensor one can use print statements, module hooks, pdb stepping etc. keeping the debugging process identical to a regular model/function execution. Further the user can print the DataParallelTensor to see the values for the tensor or its gradients for all devices at once. This allows us to see the gradient values before calling all reduce on them to investigate them individually. 

2. Composability: The DataParallelTensor composes with existing PyTorch features (thanks to torch.__dispatch__) such as vmap, vjp, jacrev, jacfwd, hessian. This makes it possible to compute jacobians and higher order derrivatives like hessians in an easy and efficient manner using vmap enabled vectorization in data parallel mode. Please see examples for details on usage.

3. Eager Philosophy: The goal of DataParallelTensor is NOT performance. It is to enable users to apply data parallel operations seamlessly to functions and models in an eager and interpretable manner. This does not mean we do not care about performance at all, please see the performance and next steps sections to get a holistic view.

API and Utilities
Initialization: The DataParallelTensors (DPT)  supports three ways of initialization:           

1. Replicated DPT: When a single tensor is supplied it is replicated across all the devices.
2. Distributed DPT: When a list/tuple of tensors is supplied, a check is made to verify if all the elements rest on different devices, if yes, then they are just wrapped as DPT else they are scattered to different devices in order.
3. Batch-Distributed DPT: When a single tensor is supplied with a batch dimension, it is uniformly chunked across the provided batch dimension and each chunk is scattered to a different device in order.                                                                                                                                          Usage: Replicated DPTs are usually used for automatically lifting the tensor args and kwargs or model parameters to DPT. Distributed DPTs are used to warp the results of operation or to convert the factory functions to DPTs. Lastly, the Batch-Distributed DPTs are used for providing inputs to models/functions. 

Utilities: We provide simple utilities to make any model DPT compatible and to perform all-reduce once the gradients for all parameters have been populated.

Note: We need to make any tensor that requires gradients explicitly data parallel since, firstly we want the multiple uses for this tensor to be recorded in its autograd graph, secondly to have a transparent view of them, thirdly to make gradient update in-place and compatible with any optimizer.

Examples in code:

1. Making any function data parallel and calling backward: As we see it just requires making the input as DPT, the rest is taken care of. The batch is chunked along dimension 0.

2. Making a model data parallel, doing a backward pass and doing all reduce: This requires three steps, making the input DPT, converting the model parameters and buffers to DPT and lastly explicitly calling all-reduce and doing a regular model update. 

3. Computing the jacobian using vjp and vmap: Similar to earlier just need to make the input DPT, all the other arguments like weight, bias and unit vectors are made DPTs dynamically before dispatch.

4. Computing hessian using hessian api and composition of jacrev and jacfwd: Demonstrating the capability to compute second order derivatives in data parallel mode. The returned hessians are also DPTs. 

Performance features:

1. Multi-threaded dispatch: Since the torch.__dispatch__ method of DPT is used to dispatch operations to different devices, to avoid serial dispatch we maintain a threadpool for parallelizing this dispatch to primarily avoid the serializing of the caching memory allocator and kernel launches.
2. Communication collectives: For initializing the DPTs and perform all reduce we use cuda communication collectives like broadcast, scatter and add-reduce to avoid serial transfer overheads.

Limitations: 

1. Performance: Since we have a single process that controls the execution and we dynamically lift the args/kwargs/factory function generated tensors to DPTs, the eager performance is slow.
2. Scaling to multiple nodes: Currently DPT only supports single node with multiple devices, we need to think about a design that allows execution on multiple nodes with multiple devices devices in an easy to use fashion.
3. Divergent Data Dependent Control flow: Since DPT operations are follow a single path of execution and since we aggregate the results for non-tenor return types (like bool), we currently do not support data-dependent control flow.

### CODE REVIEW

File: data_parallel_tensor.py [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+        future_res: List[futures.Future] = []
+        for pos in range(cls.num_threads):
+            future_res.append(
+                cls.threadpool.submit(
+                    func,
+                    *tree_map(unwrap_with_position(pos), args),
+                    **tree_map(unwrap_with_position(pos), kwargs),
+                )
+            )
+        outs = [future_res[i].result() for i in range(cls.num_threads)]
```
Comment (anijain2305) [inline data_parallel_tensor.py:174]: Nice!

File: data_parallel_tensor.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch._C import NoneType, device
+from torch._utils import _get_all_device_indices
+from torch.cuda import comm
+from torch.utils._pytree import tree_map
+
+torch.__future__.set_overwrite_module_params_on_conversion(True)
```
Comment (anijain2305) [inline data_parallel_tensor.py:12]: Add a comment on why this is needed.

File: data_parallel_tensor.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+import concurrent.futures as futures
+
+_ = torch.manual_seed(0)
```
Comment (anijain2305) [inline data_parallel_tensor.py:15]: Remove `_ =`

File: data_parallel_tensor.py [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+            # NOTE: If the input is None, we return None
+            if elem is None:
+                return None
+            assert isinstance(elem, torch.Tensor)
+            # NOTE: For handling meta tensors, if the device of an input tensor is meta,
+            # we just return the first element in such a list/tuple
+            if elem.device == device("meta"):
+                return elem
+
+            with torch.no_grad():
```
Comment (anijain2305) [inline data_parallel_tensor.py:75]: Just for curiosity. What happens if we don't have `no_grad`?

File: data_parallel_tensor.py [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+                dpt: List[torch.Tensor] = comm.broadcast(elem, devices=cls.device_ids)
+
+        elif dpt_type == DPTensorType.distributed:
+            assert isinstance(elem, list) or isinstance(elem, tuple)
+            # NOTE: For handling None values, if any element of a list/tuple is None, we return None
+            check_none = [True if e is None else False for e in elem]
```
Comment (anijain2305) [inline data_parallel_tensor.py:81]: Nit - `check_none = [e is None for e in elem]`

File: data_parallel_tensor.py [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+            if any(check_none):
+                return None
+            check_not_tensor = [
```
Comment (anijain2305) [inline data_parallel_tensor.py:84]: similarly here - you can just do `isinstance(e, torch.Tensor) for e in elem`

File: data_parallel_tensor.py [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+
+            with torch.no_grad():
+                dpt: List[torch.Tensor] = comm.broadcast(elem, devices=cls.device_ids)
+
+        elif dpt_type == DPTensorType.distributed:
+            assert isinstance(elem, list) or isinstance(elem, tuple)
+            # NOTE: For handling None values, if any element of a list/tuple is None, we return None
+            check_none = [True if e is None else False for e in elem]
+            if any(check_none):
+                return None
```
Comment (anijain2305) [inline data_parallel_tensor.py:83]: What do you think of adding an assertion - if one is None, all should be None?

File: data_parallel_tensor.py [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+                return elem[0]
+            requires_scatter: bool = False
+            with torch.no_grad():
+                for t, d_id in zip(elem, cls.device_ids):
+                    if t.device == device("meta"):
+                        # NOTE: For handling meta tensors, if the device of any tensor in such a list/tuple is meta,
+                        # we just return the first element in such a list/tuple
+                        return elem[0]
+                    if t.device != device(d_id):
+                        requires_scatter = True
```
Comment (anijain2305) [inline data_parallel_tensor.py:99]: Add a comment when it can happen. Like `torch.ones`?

File: data_parallel_tensor.py [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+            if elem is None:
+                return None
+            assert isinstance(elem, torch.Tensor)
+            # NOTE: For handling meta tensors, if the device of an input tensor is meta,
+            # we just return the first element in such a list/tuple
+            if elem.device == device("meta"):
+                return elem
+
+            with torch.no_grad():
+                dpt: List[torch.Tensor] = comm.broadcast(elem, devices=cls.device_ids)
```
Comment (anijain2305) [inline data_parallel_tensor.py:76]: maybe replace `dpt` with `dp_tensor`?

File: data_parallel_tensor.py [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+
+        # The ouput will always be a list
+        # The list can contain tensors, bools, list of tensors or tuples of tensors or None
+        # In case of tensors we just wrap them in DPT
+        # In case of list/tuple of tensors, the corresponding elements across list/tuple are warpped
+        #  into a DPT and a list/tuple is returned respectively
+
+        def out_wrap(e, func):
+            elem_type = get_element_type(e)
+            if elem_type is NoneType:
```
Comment (anijain2305) [inline data_parallel_tensor.py:202]: How about using `isinstance(e[0], Type)` for these if conditions?

File: data_parallel_tensor.py [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+                return None
+            if elem_type == torch.Tensor:
+                return DataParallelTensor(outs, func, DPTensorType.distributed)
+            elif elem_type == list:
+                return list(
+                    DataParallelTensor(list(t), func, DPTensorType.distributed)
+                    for t in zip(*e)
```
Comment (anijain2305) [inline data_parallel_tensor.py:209]: Can you have scenarios where there is a list of list of Tensors?

**Review by anijain2305 — COMMENTED** [review]
Fantastic work. Added few comments, these are mostly style related.

File: data_parallel_tensor.py (hunk 2 above)
Comment (sanketpurandare) [reply data_parallel_tensor.py:12]: Done.

File: data_parallel_tensor.py (hunk 3 above)
Comment (sanketpurandare) [reply data_parallel_tensor.py:15]: Done.

File: data_parallel_tensor.py (hunk 4 above)
Comment (sanketpurandare) [reply data_parallel_tensor.py:75]: If the tensor being replicated has it's required grad field set then the broadcast/scatter operation will be recorded in its graph and will cause the gradients to flow to the base tenor when calling backwards. Also it cannot be released from memory since it's reference is retained now.

File: data_parallel_tensor.py (hunk 5 above)
Comment (sanketpurandare) [reply data_parallel_tensor.py:81]: Done.

File: data_parallel_tensor.py (hunk 6 above)
Comment (sanketpurandare) [reply data_parallel_tensor.py:84]: Done.

File: data_parallel_tensor.py (hunk 7 above)
Comment (sanketpurandare) [reply data_parallel_tensor.py:83]: This is a design decision. We could potentially have a DPT that supports Tensors on some devices and None for some. Need to think more on this. Adding an assertion might not be the right thing to do since it is not always guaranteed that all the values should be None. We should keep this comment open for now.

File: data_parallel_tensor.py (hunk 8 above)
Comment (sanketpurandare) [reply data_parallel_tensor.py:99]: Yes usually happens with factory functions or libraries like jacfwd(forward-mode differentiation).

Comment (sanketpurandare) [reply data_parallel_tensor.py:99]: Added a comment for this.

File: data_parallel_tensor.py (hunk 7 above)
Comment (sanketpurandare) [reply data_parallel_tensor.py:83]: The reason we just decided to return None was since some operators make boolean checks on input tensors to see if they are none and since we have a single meta-data for dpt, None values become a concern.

File: data_parallel_tensor.py (hunk 11 above)
Comment (sanketpurandare) [reply data_parallel_tensor.py:209]: No we do not have such scenarios, either a single tensor or a list/tuple of tensors is returned by all operators that return a tensor.

File: data_parallel_tensor.py (hunk 7 above)
Comment (anijain2305) [reply data_parallel_tensor.py:83]: I see. In what scenarios, do we have None on some devices, and tensors on others? Have you seen any such scenarios in your testcase?

File: data_parallel_tensor.py (hunk 9 above)
Comment (sanketpurandare) [reply data_parallel_tensor.py:76]: Done.

File: data_parallel_tensor.py (hunk 7 above)
Comment (sanketpurandare) [reply data_parallel_tensor.py:83]: So according to our discussion, I make a single check to see if the values in a list are tensors or not, if they are not tensors then we return only a single value if all of them are same else raise an error. We need to insert these checks at two places:
1. DPT Constructor: When a list/tuple of values are supplied
2. When we are wrapping the output list 
Thanks for the suggestion code looks a lot cleaner and robust now :)

**Comment by anijain2305** [comment]
@Chillee Can you merge this?

**Review by anijain2305 — APPROVED** [review]
